src/aleph/http/client.clj

```diff
@@ -487,7 +712,241 @@
+        (netty/close ch)))))
+
+
+(defn- http2-req-handler
+  "Returns a fn that takes a Ring request and returns a deferred containing a
+   Ring response."
+  [{:keys [authority ch logger pipeline-transform proxy-options ssl? raw-stream? response-buffer-size]}]
+  (let [h2-bootstrap (Http2StreamChannelBootstrap. ch)]
+    (fn [req]
+      (log/trace "http2-req-handler fired")
+
+      (let [resp (d/deferred)
+            raw-stream? (get req :raw-stream? raw-stream?)
+            req' (cond-> req
+                         ;; http2 uses :authority, not host
+                         (nil? (:authority req))
+                         (assoc :authority authority)
+
+                         ;; http2 cannot leave the path empty
+                         (nil? (:uri req))
+                         (assoc :uri "/")
+
+                         (nil? (:scheme req))
+                         (assoc :scheme (if ssl? :https :http)))]
+
+        ;; create a new outbound HTTP2 stream/channel
+        (-> (.open h2-bootstrap)
+            netty/wrap-future
+            (d/chain' (fn [^Http2StreamChannel chan]
+                        (build-http2-stream-pipeline
+                          chan
+                          resp
+                          proxy-options
+                          logger
+                          pipeline-transform
+                          raw-stream?
+                          response-buffer-size)))
+            (d/chain' (fn [^Http2StreamChannel chan]
+                        (log/debug "New outbound HTTP/2 channel ready.")
+                        (if (multipart/is-multipart? req)
+                          (let [emsg "Multipart requests are not yet supported in HTTP/2"
+                                ex (ex-info emsg {:req req})]
+                            (log/error ex emsg)
+                            (throw ex))
+                          #_(@multipart-req-preprocess (assoc req :ch chan)) ; switch to HTTP1 code for multipart
+                          (http2/send-request chan req' resp))))
+            (d/catch' (fn [^Throwable t]
+                        (log/error t "Unable to open outbound HTTP/2 stream channel")
+                        (d/error! resp t)
+                        (netty/close ch))))
+
+        resp))))
+
+(defn- client-ssl-context
+  "Returns a client SslContext, or nil if none is requested.
+   Validates the ALPN setup."
+  ^SslContext
+  [ssl? ssl-context desired-protocols insecure?]
+  (if ssl?
+    (if ssl-context
+      (let [^SslContext ssl-ctx (netty/coerce-ssl-client-context ssl-context)]
+        ;; check that ALPN is set up if HTTP/2 is requested (https://www.rfc-editor.org/rfc/rfc9113.html#section-3.3)_
+        (if (and (-> ssl-ctx
+                     (.applicationProtocolNegotiator)
+                     (.protocols)
+                     (.contains ApplicationProtocolNames/HTTP_2)
+                     not)
+                 (some #(= :http2 %) desired-protocols))
```

Comment:
Shall we make the `desired-protocols` a set and perform `(desired-protocols :http2)` instead?

Reply:
We could convert it to a set in this _particular_ spot, but in general, the desired protocols are ordered, because they indicate which protocols you most prefer. E.g., `[:http1 :http2]` means you prefer HTTP/1 over HTTP/2, and the server will only fall back to HTTP/2 if it doesn't support HTTP/1.

Reply:
One annoying caveat: there's no guarantee a server will honor your preference order. 

I was debugging something today, and got confused for a hot minute because the postman-echo.com service I've been using _always_ chooses http2 if available.

Reply:
Understood!